modules/auxiliary/scanner/scada/bacnet_l3.rb

```diff
@@ -0,0 +1,412 @@
+    "\x81\n\u0000\u0011\u0001\u0004\u0002\u0002\u0000\f\f\u0002{object_identifier}#{READ_FIRMWARE_VERSION_PROP}", # firmware-revision
+    "\x81\n\u0000\u0011\u0001\u0004\u0002\u0002\u0000\f\f\u0002{object_identifier}#{READ_APP_SOFT_VERSION_PROP}", # application-software-version
+    "\x81\n\u0000\u0011\u0001\u0004\u0002\u0002\u0000\f\f\u0002{object_identifier}#{READ_DESCRIPTION_PROP}"
+  ].freeze # description
+
+  GET_PROPERTY_MESSAGES_L3_NESTED = [
+    "\u0001${dest_net_id}{dadr_len}{dadr}\xFF\u0002\u0002\u0002\f\f\u0002{object_identifier}#{READ_MODEL_NAME_PROP}",
+    "\u0001${dest_net_id}{dadr_len}{dadr}\xFF\u0002\u0002\u0002\f\f\u0002{object_identifier}#{READ_FIRMWARE_VERSION_PROP}",
+    "\u0001${dest_net_id}{dadr_len}{dadr}\xFF\u0002\u0002\u0002\f\f\u0002{object_identifier}#{READ_APP_SOFT_VERSION_PROP}",
+    "\u0001${dest_net_id}{dadr_len}{dadr}\xFF\u0002\u0002\u0002\f\f\u0002{object_identifier}#{READ_DESCRIPTION_PROP}"
+  ].freeze
+
+  # Global variables initialization
+  $iface
+  $timeout
+  $packet_count
+  $port
+
+  def initialize
+    super(
+      'Name' => 'BACnet Scanner',
+      'Description' => '
+        Discover BACnet devices by broadcasting Who-is message, then poll
+        discovered devices for properties including model name,
+        software version, firmware revision and description.
+      ',
+      'Author' => ['Paz @ SCADAfence'],
+      'License' => MSF_LICENSE,
+      'Notes' => {
+        'Stability' => [CRASH_SAFE],
+        'Reliability' => [UNRELIABLE_SESSION],
+        'SideEffects' => [SCREEN_EFFECTS]
+      }
+    )
+
+    register_options(
+      [
+        OptInt.new('TIMEOUT', Array[true, 'The socket connect timeout in seconds', DEFAULT_SERVER_TIMEOUT]),
+        OptInt.new('COUNT', Array[true, 'The number of times to send each packet', DEFAULT_SEND_COUNT]),
+        OptPort.new('PORT', Array[true, 'BACnet/IP UDP port to scan (usually between 47808-47817)', DEFAULT_BACNET_PORT]),
+        OptString.new('INTERFACE', Array[true, 'The interface to scan from', 'eth1'])
+      ], self.class
+    )
+  end
+
+  def hex_to_bin(s)
+    s.scan(/../).map { |x| x.hex.chr }.join
+  end
+
+  def bin_to_hex(s)
+    s.each_byte.map { |b| b.to_s(16).rjust(2, '0') }.join
+  end
+
+  def to_ip_address(s)
+    s = s.unpack('C*')
+    "#{s[0]}.#{s[1]}.#{s[2]}.#{s[3]}"
+  end
+
+  # Create UDP packet with the l2 and l3 details.
+  def create_packet_base
+    udp_pkt = PacketFu::UDPPacket.new
+    udp_pkt.ip_daddr = '255.255.255.255'
+    udp_pkt.eth_daddr = 'ff:ff:ff:ff:ff:ff'
+    udp_pkt.udp_dport = $port
+    udp_pkt.udp_sport = rand(0xffff - 1024) + 1024
+    udp_pkt.ip_ttl = 64
+
+    begin
+      udp_pkt.ip_saddr = `ip a s #{$iface} | grep "inet "`.scan(%r{inet (.*)/}).flatten[0] # Get interface's ip address
+      udp_pkt.eth_saddr = `ip a s #{$iface} | grep ether`.scan(%r{link/ether (.*) brd}).flatten[0] # Get interface's mac address
```

Comment:
You can include the `Msf::Exploit::Capture` [mixin](https://github.com/rapid7/metasploit-framework/blob/master/lib/msf/core/exploit/capture.rb) that brings many options and useful helpers such as `get_mac` to retrieve the MAC address.

Reply:
Fixed